lib/assets/chart_builder/main.js

```diff
@@ -0,0 +1,301 @@
+
+  function handleFieldChange(event) {
+    const { name, value } = event.target;
+    if (channelFields.includes(name)) { maybeReset(name, value) };
+    ctx.pushEvent("update_field", { field: name, value });
+  }
+
+  function updateDataOptions(event) {
+    const { value } = event.target;
+    const options = state.data_options[value];
+    setAxisOptions(options);
+  }
+
+  ctx.handleEvent("update", ({ fields }) => {
+    clearUnavailableData();
+    setValues(fields);
+  });
+
+  ctx.handleEvent("missing_dep", ({ dep }) => {
+    updateInfoBox(dep);
+  });
+
+  ctx.handleEvent("set_available_data", ({ options }) => {
+    state.data_options = options;
+    Object.keys(options).length ? setAvailableData(options) : updateDataInfoBox();
+  });
+
+  function setValues(fields) {
+    for (const field in fields) {
+      ctx.root.querySelector(`[name="${field}"]`).value = fields[field];
+    }
+  }
+
+  function setAvailableData(options) {
+    enableAll();
+    const selectedData = selectDataElem.value;
+    if (!selectedData) {return setFreshData(options)};
+    const availableDataOptions = options[selectedData];
+    updateDataInfoBox(selectedData);
+    selectDataElem.innerHTML = "";
+    createOptions(selectDataElem, Object.keys(options));
+    selectDataElem.value = selectedData;
+    if (availableDataOptions) {
+      setAvailableDataOptions(availableDataOptions);
+      selectDataElem.classList.remove("unavailable");
+    } else {
+      createOptions(selectDataElem, [selectedData], "unavailable-option");
+      selectDataElem.value = selectedData;
+      disableAll();
+      selectDataElem.disabled = false;
+      selectDataElem.classList.add("unavailable");
+    }
+  }
+
+  function renderCurrentData(fields) {
+    disableAll();
+    for (const field of channelFields.concat(["data_variable"])) {
+      const elem = ctx.root.querySelector(`[name="${field}"]`);
+      elem.innerHTML = `<option value="${fields[field]}">${fields[field]}</option>`;
+    }
+    setValues(fields);
+    updateDataInfoBox(fields["data_variable"]);
+  }
+
+  function setFreshData(options) {
+    enableAll();
+    selectDataElem.innerHTML = "";
+    createOptions(selectDataElem, Object.keys(options));
+    const selected = selectDataElem.value;
+    const axisOptions = options[selected];
+    updateDataInfoBox(options[selected]);
+    if (axisOptions) { setAxisOptions(axisOptions) };
+    ctx.pushEvent("update_field", { field: "data_variable", value: selected });
+  }
+
+  function setAxisOptions(options) {
+    enableAll();
+    for (const elem of channelFields) {
+      const selectElem = ctx.root.querySelector(`[name="${elem}"]`);
+      const inner = elem === "color_field" ? "<option value=''></option>" : "";
+      selectElem.innerHTML = inner;
+      createOptions(selectElem, options);
+      const selected = selectElem.value;
+      ctx.pushEvent("update_field", { field: elem, value: selected });
+    }
+    setTypes();
```

Comment:
We are sending `"update_field"` in the loop above, and then in `setTypes` in a loop as well. We shouldn't be spamming the server so much.

This bulk change is a result of selecting the dataframe, to set the new defaults we can either:

1. Compute the new values on the client and send as multi-field update, then broadcast all those fields
2. Send just the new dataframe name and determine the new values on the server, then broadcast all those fields

I think 2. is more direct and will end up better :)